### PR TITLE
Create Authentication for CFSSL API -- WIP

### PR DESCRIPTION
**Pull request by wl2002**
Created a Verify method to make sure that a JWT was of the valid form. It seems that to verify JWTs it requires a public key, so I'm not entirely sure how that will work in production. I'm also trying to figure out how you would make the verification general enough to look at a token's claims when in production the claim values could be set to anything.

### CODE REVIEW

File: jwtauth/jwtauth.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Package jwtauth implements an interface for providing CFSSL
+// authentication.
+package jwtauth
+
+import (
+	//"flag"
```
Comment (kisom) [inline jwtauth/jwtauth.go]: Should remove commented out imports (or just run `goimports` on every save to automatically add/remove imports).

**Comment by grittygrease** [comment]
This needs to be integrated as a authentication provider in the "auth" package: https://github.com/cloudflare/cfssl/blob/master/auth/auth.go

Furthermore, this form of authentication should only be available over https.

**Comment by grittygrease** [comment]
Usually public keys are obtained from a known URL. It's worth creating these options in the auth section of the configuration file, including a URL for where to obtain the public key.

**Comment by dqminh** [comment]
> Usually public keys are obtained from a known URL. It's worth creating these options in the auth section of the configuration file, including a URL for where to obtain the public key.

Agreed, the verification code should have logic to fetch the public keys and cache it until it expires.

https://github.com/coreos/go-oidc implemented some nice functions to do this, for example:
- https://github.com/coreos/go-oidc/blob/e9c08071fd4b03b7c62fa47742350dc4c85505ae/oidc/verification.go
- https://github.com/coreos/go-oidc/blob/ee7cb1fb480df22f7d8c4c90199e438e454ca3b6/oidc/key.go#L14

Can we reuse them in cfssl ?

**Comment by wl2002** [comment]
I'll be moving these to the auth package and will look at those functions to see how to integrate them into cfssl in the next commit.

File: jwtauth/jwtauth.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	Request       []byte `json:"request"`
+}
+*/
+// New attempts to create a new JWT based off of the signing key passed into it
+func New(mySigningKey []byte) (string, error) {
+
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Claims["exp"] = time.Now().Add(time.Hour * 72).Unix()
+	tokenString, err := token.SignedString(mySigningKey)
+	return tokenString, err
```
Comment (j-delaney) [inline jwtauth/jwtauth.go]: You could just remove the line above this one and do `return token.SignedString(mySigningKey)`

File: jwtauth/testdata/token [hunk 3]
```diff
@@ -0,0 +1 @@
+eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.EkN-DOsnsuRjRO6BxXemmJDm3HbxrbRzXglbN2S4sOkopdU4IsDxTI8jO19W_A4K8ZPJijNLis4EZsHeY559a4DFOd50_OqgHGuERTqYZyuhtF39yxJPAjUESwxk2J5k_4zM3O-vtd1Ghyo4IbqKKSy6J9mTniYJPenn5-HIirE
```
Comment (j-delaney) [inline jwtauth/testdata/token]: Missing newline at end of file

File: jwtauth/testdata/token.pubkey [hunk 4]
```diff
@@ -0,0 +1,3 @@
+-----BEGIN PUBLIC KEY-----
+MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDdlatRjRjogo3WojgGHFHYLugdUWAY9iR3fy4arWNA1KoS8kVw33cJibXr8bvwUAUparCwlvdbH6dvEOfou0/gCFQsHUfQrSDv+MuSUMAe8jzKE4qW+jK+xQU9a03GUnKHkkle+Q0pX/g6jXZ7r1/xAK5Do2kQ+X5xK9cipRgEKwIDAQAB
+-----END PUBLIC KEY-----
```
Comment (j-delaney) [inline jwtauth/testdata/token.pubkey]: Missing newline at end of file

File: jwtauth/jwtauth.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+func Verify(pubkey []byte, myToken string) (bool, error) {
+	token, err := jwt.Parse(string(myToken), func(t *jwt.Token) (interface{}, error) {
+		return pubkey, nil
+	})
+
+	if err != nil {
+		return false, err
+	}
+	if token.Valid {
+		fmt.Println("claims == ", token.Claims)
```
Comment (codyli520) [inline jwtauth/jwtauth.go]: Debug code?

File: jwtauth/jwtauth.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+type AuthenticatedRequest struct {
+	// An Authenticator decides whether to use this field.
+	Timestamp     int64  `json:"timestamp,omitempty"`
+	RemoteAddress []byte `json:"remote_address,omitempty"`
+	Token         []byte `json:"token"`
+	Request       []byte `json:"request"`
+}
+*/
```
Comment (codyli520) [inline jwtauth/jwtauth.go]: Looks like Travis is treating this giant block of commented-out code as part of the comment for exported function New() below. Consider remove this block and keep it somewhere for your own reference if you are not sure at the moment.

**Comment by lziest** [comment]
orphaned PR.

**Pull request closed without merging**